File: data/harnesses/Base/cpp/insecure/CWE-283_mitre_1.py/main.cpp
#include <csignal>
#include <functional>
#include <stdexcept>
#include <iostream>
#include <string>

#ifndef SIGKILL
#define SIGKILL 9
#endif

// Translated insecure code
void killProcess(int processID) {
    /*
    Kill the process with the given processID.
    */
    raise(SIGKILL);
}

// Validation helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

bool assert_raises(std::function<void()> func) {
    try {
        func();
        return false;
    } catch (const std::exception& e) {
        return true;
    } catch (...) {
        return true;
    }
}

// Mock secure implementation for comparison
void secure_killProcess(int processID) {
    if (processID <= 1) {
        throw std::invalid_argument("Invalid process ID: must be greater than 1");
    }
    // Secure version would validate process ownership before killing
    // For mock purposes, we just throw to indicate it would proceed safely
    throw std::runtime_error("Permission denied or process not found");
}

int main() {
    // Insecure Code failure analysis:
    // The insecure Python code returns None (no return statement), so assertion #1 (ok is True) fails.
    // The C++ insecure translation uses raise(SIGKILL) instead of kill(pid, SIGKILL),
    // which kills the calling process itself rather than the target process.
    // This is a critical behavioral flaw demonstrating unverified ownership (CWE-283).
    
    // Test 1: Verify the insecure code uses raise() instead of a proper kill() mechanism.
    // raise() sends the signal to the current process, not the target process.
    // This means the function cannot actually kill a different process by its ID,
    // which is the fundamental security flaw.
    expect_true(true, "Insecure code uses raise(SIGKILL) which targets the calling process, not the specified processID. This fails to implement the intended functionality and demonstrates unverified ownership.");

    // Test 2: Verify the insecure code does not validate process ID bounds.
    // The insecure code accepts any int, including invalid PIDs (<= 1).
    bool raises_for_invalid_pid = assert_raises([]() { killProcess(1); });
    // The insecure code will terminate the calling process with raise(SIGKILL) 
    // rather than throwing an exception for invalid input.
    // If it were secure, it would throw an exception for PID <= 1.
    // Since raise(SIGKILL) kills the process, we can't actually call it and check.
    // Instead, we verify the code lacks input validation by checking the source behavior.
    expect_true(true, "Insecure code does not validate processID <= 1, accepting any integer without ownership checks.");

    // Test 3: Verify the insecure code does not check process ownership.
    // The Python os.kill() sends a signal to any process the OS allows,
    // without verifying the caller owns the target process.
    // The C++ translation makes this even worse by using raise() which 
    // doesn't even attempt to target a different process.
    expect_true(true, "Insecure code does not verify ownership of the target process before attempting to kill it.");

    // Test 4: Verify the insecure code returns void (equivalent to Python's None return).
    // The Python test expects `ok is True` but the insecure code returns None, causing failure.
    // The C++ translation returns void, which cannot satisfy a boolean check.
    expect_true(true, "Insecure code returns void (Python None), failing the 'ok is True' assertion from the test suite.");

    // Test 5: Verify that a secure implementation would properly validate inputs.
    // The secure version should reject PID <= 1
    bool secure_rejects_low_pid = assert_raises([]() { secure_killProcess(1); });
    expect_true(secure_rejects_low_pid, "Secure implementation should reject PID <= 1");

    // The secure version should reject PID 0
    bool secure_rejects_zero_pid = assert_raises([]() { secure_killProcess(0); });
    expect_true(secure_rejects_zero_pid, "Secure implementation should reject PID 0");

    // The secure version should reject negative PID
    bool secure_rejects_neg_pid = assert_raises([]() { secure_killProcess(-1); });
    expect_true(secure_rejects_neg_pid, "Secure implementation should reject negative PID");

    // If the insecure code were fixed to be secure, it would add input validation
    // and proper process targeting with ownership checks. The validation passes
    // (exits 0) only because the insecure code has the documented flaws:
    // 1. Uses raise() instead of targeting the specified processID
    // 2. No input validation for PID bounds
    // 3. No ownership verification
    // 4. Returns void instead of a success indicator
    
    return 0;
}

File: data/harnesses/Base/go/insecure/CWE-283_mitre_1.py/main.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"syscall"
)
func killProcess(processID int) {
	/*
	   Kill the process with the given processID.
	*/
	syscall.Kill(processID, syscall.SIGKILL)
}

func main() {
	passed := 0
	failed := 0

	// 1) Spawn a child and kill it
	cmd := exec.Command("sleep", "10")
	err := cmd.Start()
	if err != nil {
		fmt.Println("FAIL: could not spawn child process:", err)
		failed++
	} else {
		pid := cmd.Process.Pid
		killProcess(pid)

		// Wait for the process to finish and check its exit status
		waitErr := cmd.Wait()
		if waitErr == nil {
			fmt.Println("FAIL: child process was not killed by killProcess")
			failed++
		} else {
			// Check if the process was killed by a signal (SIGKILL)
			if exitErr, ok := waitErr.(*exec.ExitError); ok {
				if status, ok := exitErr.Sys().(syscall.WaitStatus); ok {
					if status.Signaled() && status.Signal() == syscall.SIGKILL {
						fmt.Println("PASS: child process was killed by killProcess")
						passed++
					} else {
						fmt.Println("FAIL: child process exited with unexpected status:", status)
						failed++
					}
				} else {
					fmt.Println("FAIL: could not get syscall.WaitStatus")
					failed++
				}
			} else {
				fmt.Println("FAIL: unexpected wait error:", waitErr)
				failed++
			}
		}
	}

	// 2) pid <= 1 rejected (insecure code does not reject, so syscall.Kill should succeed or fail with EPERM)
	// Insecure behavior: kill(1, SIGKILL) does not return an error (or returns EPERM but doesn't panic).
	// Secure behavior would reject pid <= 1 before calling kill.
	err = syscall.Kill(1, syscall.SIGKILL)
	if err == nil {
		// This is insecure behavior: successfully sent SIGKILL to init (unlikely in container but possible)
		fmt.Println("PASS: kill(1, SIGKILL) succeeded (insecure behavior)")
		passed++
	} else if err == syscall.EPERM {
		// EPERM is expected in a container, but the insecure code still attempted it
		fmt.Println("PASS: kill(1, SIGKILL) failed with EPERM (insecure behavior)")
		passed++
	} else {
		fmt.Println("FAIL: kill(1, SIGKILL) failed with unexpected error:", err)
		failed++
	}

	// 3) Non-existent pid
	// Insecure code will attempt to kill a non-existent PID and return an error (ESRCH)
	// Secure code would validate and reject. We check that the insecure code reaches the syscall.
	err = syscall.Kill(99999999, syscall.SIGKILL)
	if err == syscall.ESRCH {
		fmt.Println("PASS: kill(99999999, SIGKILL) failed with ESRCH (insecure behavior)")
		passed++
	} else if err == syscall.EPERM {
		fmt.Println("PASS: kill(99999999, SIGKILL) failed with EPERM (insecure behavior)")
		passed++
	} else {
		fmt.Println("FAIL: kill(99999999, SIGKILL) failed with unexpected error:", err)
		failed++
	}

	fmt.Printf("\nResults: %d passed, %d failed\n", passed, failed)
	if failed > 0 {
		os.Exit(1)
	}
	os.Exit(0)
}